pages.ts

```diff
@@ -83,3 +94,140 @@ export async function loadData(
+  siteId: number
+): Promise<PageWithParams | null> => {
+  const { data: pages, error } = await getSupabaseClient()
+    .from("pages")
+    .select("id, name, data, path")
+    .match({ site: siteId, archived: false })
+    .is("flag", null)
+    .is("archived", false);
+
+  const routes = pages?.map((page) => ({
+    page,
+    pattern: page.path,
+  })) as Array<{ pattern: string; page: Page }>;
+
+  sortRoutes(routes);
+
+  const matchRoute = routes
+    .map(({ pattern, page }) => {
+      const urlPattern = new URLPattern({ pathname: pattern });
+      const result = urlPattern.exec({ pathname: path });
+
+      return { match: !!result, params: result?.pathname.groups, page };
+    })
+    .find(({ match }) => match);
+
+  if (error || !matchRoute) {
+    console.error(error?.message || `Page with path "${path}" not found`);
+
+    return null;
+  }
+
+  return matchRoute;
+};
+
+/**
+ * Fetchs a specific page from the database and also
+ * computes the page params base on the request's URL if provided
+ */
+export const fetchPageFromId = async (
+  pageId: number,
+  pathname?: string
+): Promise<PageWithParams> => {
+  const { data: pages, error } = await getSupabaseClient()
+    .from("pages")
+    .select("id, name, data, path")
+    .match({ id: pageId });
+
+  const matchPage = pages?.[0];
+
+  if (error || !matchPage) {
+    throw new Error(error?.message || `Page with id ${pageId} not found`);
+  }
+
+  const urlPattern = new URLPattern(matchPage.path);
+  const params = pathname
+    ? urlPattern.exec(pathname)?.pathname.groups
+    : undefined;
+
+  return {
+    page: matchPage as Page,
+    params,
+  };
+};
+
+/**
+ * Fetches a page containing this component.
+ *
+ * This is used for creating the canvas. It retrieves
+ * or generates a fake page from the database at
+ * /_live/components/<componentName.tsx>
+ *
+ * This way we can use the page editor to edit components too
+ */
+export const fetchPageFromComponent = async (
+  component: string // Ex: Banner.tsx
+): Promise<PageWithParams> => {
+  const supabase = getSupabaseClient();
+  const { component: instance, loaders } = createComponent(
+    `./components/${component}`
+  );
+  const page = createPageForComponent(component, {
+    components: [instance],
+    loaders,
+  });
+
+  if (!exists(`./components/${component}`)) {
+    throw new Error(`Component at ${component} Not Found`);
+  }
+
+  const { data } = await supabase
+    .from<Page>("pages")
+    .select("id, path")
+    .match({ path: page.path });
+
+  const match = data?.[0];
+
+  if (match) {
+    return fetchPageFromId(match.id);
+  }
+
+  return { page };
+};
+
+/**
+ * Sort pages by their relative routing priority, based on the parts in the
+ * route matcher
+ *
+ * Extracted from: https://github.com/denoland/fresh/blob/046fcde959041ac9cd5f2b39671c819c4af5cc24/src/server/context.ts#L683
+ */
+export function sortRoutes<T extends { pattern: string }>(routes: T[]) {
+  routes.sort((a, b) => {
```

Comment:
A small nit.

Create a rankRoute function that given a route, returns an integer and use this function to rank both parameters on the sort function. 

```suggestion
  routes.sort((a, b) => rankRoute(a) - rankRoute(b))

```

Reply:
LGTY?

![image](https://user-images.githubusercontent.com/18706156/199750628-16bcea88-517e-4452-a659-542ba9ceaa8b.png)

Reply:
perfect!